### PR TITLE
Go lambda instrumentation enhancements

### PR DESCRIPTION
**Pull request by bhautikpip**
1. This PR introduces Go Lambda instrumentation enhancements to make `xrayconfig` package more useful.

### CODE REVIEW

File: instrumentation/github.com/aws/aws-lambda-go/otellambda/xrayconfig/xrayconfig.go
Comment (bhautikpip) [reply instrumentation/github.com/aws/aws-lambda-go/otellambda/xrayconfig/xrayconfig.go:63]: We have followed that pattern in `config.go` file in `otellambda` package. I think for `xrayconfig.go` we have followed naming without `With` keyword. See [here](https://github.com/open-telemetry/opentelemetry-go-contrib/blob/main/instrumentation/github.com/aws/aws-lambda-go/otellambda/xrayconfig/xrayconfig.go#L91) for `Propagator` and `EventToCarrier` method name for reference.

File: instrumentation/github.com/aws/aws-lambda-go/otellambda/xrayconfig/README.md
Comment (bhautikpip) [reply instrumentation/github.com/aws/aws-lambda-go/otellambda/xrayconfig/README.md:49]: Looks like `InstrumentHandler` is the correct name. We have changed the name in original merge of lambda instrumentation to address reviewer feedback. Ref: https://github.com/open-telemetry/opentelemetry-go-contrib/blob/main/instrumentation/github.com/aws/aws-lambda-go/otellambda/wrapLambdaHandler.go#L98

File: instrumentation/github.com/aws/aws-lambda-go/otellambda/xrayconfig/README.md
Comment (bhautikpip) [reply instrumentation/github.com/aws/aws-lambda-go/otellambda/xrayconfig/README.md:37]: yep. see this [comment](https://github.com/open-telemetry/opentelemetry-go-contrib/pull/1413#discussion_r745193392)

File: instrumentation/github.com/aws/aws-lambda-go/otellambda/xrayconfig/xrayconfig.go
Comment (Aneurysm9) [reply instrumentation/github.com/aws/aws-lambda-go/otellambda/xrayconfig/xrayconfig.go:63]: I understand that other functions in the `xrayconfig` package were not adhering to the convention.  I now think that to be a mistake that we should fix at the earliest opportunity, a.k.a., now.

File: instrumentation/github.com/aws/aws-lambda-go/otellambda/lambda.go
Comment (Aneurysm9) [inline instrumentation/github.com/aws/aws-lambda-go/otellambda/lambda.go:30]:
```suggestion
```

This is unnecessary and must be removed.  Instrumentation packages must never depend on an SDK implementation.  The instrumentation defines its own `Flusher` interface so that it does not need to explicitly depend on any concrete type.

Comment (bhautikpip) [reply instrumentation/github.com/aws/aws-lambda-go/otellambda/lambda.go:30]: If we remove this then we might need to add this method (https://github.com/open-telemetry/opentelemetry-go-contrib/pull/1413#discussion_r745082701) again right? I tested the code without it and we're not seeing instrumentation nodes coming up properly.

File: instrumentation/github.com/aws/aws-lambda-go/otellambda/xrayconfig/xrayconfig.go
Comment (bhautikpip) [reply instrumentation/github.com/aws/aws-lambda-go/otellambda/xrayconfig/xrayconfig.go:63]: I think it would be confusing because then public methods in `otellambda` packages would have the same name as public methods in `xrayconfig` packages and would confuse users but I'm open do it like the way you mentioned but just posting this as a concern.

File: instrumentation/github.com/aws/aws-lambda-go/otellambda/lambda.go
Comment (Aneurysm9) [reply instrumentation/github.com/aws/aws-lambda-go/otellambda/lambda.go:30]: No, there is no need for that.  This won't be released until after the next SDK release.  The `go.mod` files that need to import the SDK will be updated as appropriate at that time.  We do not make any guarantees regarding the ability to use unreleased commits of the SDK or instrumentation.

File: instrumentation/github.com/aws/aws-lambda-go/otellambda/xrayconfig/xrayconfig.go
Comment (Aneurysm9) [reply instrumentation/github.com/aws/aws-lambda-go/otellambda/xrayconfig/xrayconfig.go:63]: We should simply remove the superfluous functions.  The `TracerProvider()` and `Flusher()` functions provide no value as they merely pass through to the functions in the instrumentation package.  `WithRecommendedOptions()` can call the `Option`-generating functions in the instrumentation package directly.  `Propagator()` and `EventToCarrier()` should be renamed with the `With` prefix.

**Review by Aneurysm9 — CHANGES_REQUESTED** [review]
Instrumentation must not depend on an SDK.

File: instrumentation/github.com/aws/aws-lambda-go/otellambda/xrayconfig/README.md
Comment (Aneurysm9) [inline instrumentation/github.com/aws/aws-lambda-go/otellambda/xrayconfig/README.md:59]:
```suggestion
| `WithEventToCarrier` | Function which reads X-Ray TraceID from Lambda environment and inserts it into a `propagtation.TextMapCarrier` | Individually exported as `WithEventToCarrier()`, also included in `WithRecommendedOptions()`
| `WithPropagator` | An `xray.propagator` | Individually exported as `WithPropagator()`, also included in `WithRecommendedOptions()`
```

File: instrumentation/github.com/aws/aws-lambda-go/otellambda/xrayconfig/xrayconfig.go
Comment (Aneurysm9) [inline instrumentation/github.com/aws/aws-lambda-go/otellambda/xrayconfig/xrayconfig.go:61]:
```suggestion
// WithEventToCarrier returns an otellambda.Option to enable
```

File: instrumentation/github.com/aws/aws-lambda-go/otellambda/xrayconfig/xrayconfig.go
Comment (Aneurysm9) [inline instrumentation/github.com/aws/aws-lambda-go/otellambda/xrayconfig/xrayconfig.go:69]:
```suggestion
// WithPropagator returns an otellambda.Option to enable the xray.Propagator
```

**Comment by Aneurysm9** [comment]
The failing unit test is a result of no longer trying to yield to the go runtime before invoking `ForceFlush()`.  Once https://github.com/open-telemetry/opentelemetry-go/pull/2335 is available in a release then the test should start passing and this PR can proceed.

**Comment by bhautikpip** [comment]
@MrAlias @dashpole Can we review this change? So that when we release the core this should be ready to merge.

**Review by Aneurysm9 — APPROVED** [review]

**Comment by alolita** [comment]
@open-telemetry/proto-go-approvers @dashpole can you please provide a second review so that this PR can be merged. We have a downstream dependency on this. Many thanks in advance.

File: instrumentation/github.com/aws/aws-lambda-go/otellambda/xrayconfig/xrayconfig.go
Comment (dashpole) [inline instrumentation/github.com/aws/aws-lambda-go/otellambda/xrayconfig/xrayconfig.go:43]: We probably don't need to log here.  Since we return the error, we expect the caller to handle it (and log it if they would like).  Same applies below and elsewhere.

File: instrumentation/github.com/aws/aws-lambda-go/otellambda/xrayconfig/xrayconfig_test.go
Comment (dashpole) [inline instrumentation/github.com/aws/aws-lambda-go/otellambda/xrayconfig/xrayconfig_test.go:158]: Probably best to check that the err is non-nil here.

**Review by dashpole — APPROVED** [review]
Just nits.  Looks good overall

File: instrumentation/github.com/aws/aws-lambda-go/otellambda/xrayconfig/xrayconfig_test.go
Comment (Aneurysm9) [inline instrumentation/github.com/aws/aws-lambda-go/otellambda/xrayconfig/xrayconfig_test.go:161]:
```suggestion
	assert.NoError(t, err)
```

**Pull request merged**
